homeassistant/components/kodi/binary_sensor.py

```diff
@@ -0,0 +1,165 @@
+        """Switch sensor off from api call."""
+        self._is_on = False
+        self.async_write_ha_state()
+        _LOGGER.debug("Kodi %s off (%s)", self.name, self._api_off)
+
+    @callback
+    def async_dummy(self, sender, data):
+        """Empty function to replace websocket callbacks on lost/closed connection."""
+
+    async def async_added_to_hass(self) -> None:
+        """Register callbacks for needed api endpoints."""
+        await self._connman.add_callback_on_connect(self._on_ws_connected)
+        await self._connman.add_callback_on_disconnect(self._on_ws_disconnected)
+
+    @callback
+    async def _on_ws_connected(self):
+        """Call after ws is connected."""
+        _LOGGER.debug("Binary sensor %s websocket connected", self.name)
+        await self.async_query_display_status()
+        await self._register_ws_callbacks()
+
+    @callback
+    async def _on_ws_disconnected(self):
+        """Call after ws is connected."""
+        await self._unregister_ws_callbacks()
+        await self._clear_connection()
+
+    @callback
+    async def _register_ws_callbacks(self):
+        """Register kodi websocked callbacks for this sensor."""
+        _LOGGER.debug("Setting up binary sensor callbacks for Kodi %s", self.name)
+        server = self._connman.connection.server
+        setattr(server, self._api_on, self.async_on)
+        setattr(server, self._api_off, self.async_off)
+
+    @callback
+    async def _unregister_ws_callbacks(self):
+        """Register kodi websocked callbacks for this sensor."""
+        _LOGGER.debug("Removing binary sensor callbacks for %s", self.name)
+        server = self._connman.connection.server
+        setattr(server, self._api_on, self.async_dummy)
+        setattr(server, self._api_off, self.async_dummy)
+
+    async def async_query_display_status(self):
+        """Retrieve boolean Kodi display properties."""
+        display_properties = {"booleans": [self._kodi_boolean]}
+        display_status = await self._connman.kodi.call_method(
+            "XBMC.GetInfoBooleans", **display_properties
+        )
+        self._is_on = display_status[self._kodi_boolean]
+        _LOGGER.debug("Query statue: %d %s", self._is_on, display_status)
+        self.async_write_ha_state()
+
+    async def _clear_connection(self):
+        _LOGGER.debug("Binary sensor %s clear connection", self.name)
+        self._reset_state()
+        self.async_write_ha_state()
+
+    def _reset_state(self):
+        """Set state to default."""
+        self._is_on = False
+
+    async def async_update(self) -> None:
+        """Retrieve latest state."""
+        if not self._connman.connected:
+            self._reset_state()
+            return
+
+        try:
+            await self.async_query_display_status()
+        except TransportError:
+            self._reset_state()
+            return
+
+    @property
+    def should_poll(self) -> bool:
```

Comment:
Do you think this, and a few other things, can be extracted to a common base class for the binary sensor and media player?

Reply:
I could see there:
* `_attr_has_entity_name = True`
* `def should_poll(self) -> bool:`
* maybe `_connman`

and maybe the abstract for:
* `async def async_update(self) -> None:`

to force the implementation of "poll". I'm not sure if the cost-benefit ratio is enough. Or do you see more?

Reply:
I couldn't resist doing it also for me to learn.

Reply:
I think a lot of the callback code can probably be extracted too

Reply:
The callback registering is mostly platform internal. I didn't want to force too many abstract methods.